### ISSUE TITLE
Super cool work

### ISSUE THREAD

**Issue by christophschuhmann** [jordandare/echo-tts#1]
Super cool work!

Could you release the training code? would love to tune multilingual versions and add more emotion control.

**Comment by humair-m** [comment]
Good  question

**Comment by jordandare** [comment]
Thanks so much for your interest! The training code is in JAX for TPU. I can share the "core" train step code.

```
def train_step(
    state: train_state.TrainState,
    batch: PyTreeDef,
    rng: jax.Array,
    is_eval: bool = False,
) -> Tuple[train_state.TrainState, Dict] | Tuple[jax.Array, Dict]:
    
    rng_sigma, rng_sigma_mod, rng_speaker_cfg, rng_x = jax.random.split(rng, 4)


    def loss_fn(
        params: PyTreeDef,
    ) -> Tuple[jax.Array, Dict]:
        
        # normal_sample = jax.random.normal(rng_sigma, (batch['latent'].shape[0], ))
        # t = jax.lax.logistic(normal_sample)


        # for stratified sampling, the offset/mod is needed in order to avoid a subtle bug in data-host-distributed setups (restricting data to certain timestep ranges)!
        batch_size = batch['latent'].shape[0]
        offset = jax.random.uniform(rng_sigma, ())
        mod_offset = jax.random.randint(rng_sigma_mod, (), 0, batch_size)
        quantiles = ((jnp.arange(batch_size, dtype=jnp.int32) + mod_offset) % batch_size + offset) / batch_size
        t = jax.lax.logistic(
            jax.scipy.stats.norm.ppf(quantiles)
        )

        speaker_cfg_mask = jax.random.bernoulli(rng_speaker_cfg, 0.9, batch_size)
        # text cfg done outside of this loop


        noise = jax.random.normal(rng_x, batch['latent'].shape)
        x_noised = batch['latent'] * (1 - t[..., None, None]) + noise * t[..., None, None]
        
        model_out = model.apply(
            {'params': params}, 
            x=x_noised,
            t=t, 
            text_input_ids=batch['text_input_ids'], 
            text_mask=batch['text_mask'],
            speaker_latent=batch['speaker_latent'] * speaker_cfg_mask[..., None, None],
            speaker_mask=batch['speaker_latent_mask'] * speaker_cfg_mask[..., None],
        )
        
        diffusion_loss = (model_out - (noise - batch['latent'])) ** 2

        diffusion_loss = jnp.mean(diffusion_loss) / (jnp.mean(batch['latent_mask']) + 1e-6) # normalize by non-padding but still predict non-padding
        
        loss = diffusion_loss

        metrics = {'train/loss': loss}

        metrics['train/average_latent_mask'] = jnp.mean(batch['latent_mask'])
        metrics['train/average_speaker_latent_mask'] = jnp.mean(batch['speaker_latent_mask'])

        return loss, metrics

    if is_eval:
        return loss_fn(state.params)

    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
    (_, metrics), grads = grad_fn(state.params)

    metrics['train_other/raw_grad_norm'] = l2_norm(grads)
    state = state.apply_gradients(grads=grads)

    metrics['train_other/param_norm'] = l2_norm(state.params)
    metrics['train_other/lr'] = lr_schedule(state.step) if callable(lr_schedule) else lr_schedule

    return state, metrics
```

The model is trained with the RF objective (with logit-normal timestep sampling), so in PyTorch, the pseudocode (given something like the current, inference-oriented model code) would look vaguely like:

```
speaker_latent, speaker_mask = ae_encode_with_mask(speaker_latent_up_to_5_minutes_but_can_be_less) # get_speaker_latent_and_mask
latent = ae_encode(audio_30seconds) # (including pca transformation / scaling)
text_input_ids, text_mask = get_text_input_ids_and_mask(text)

speaker_kv = get_kv_cache_speaker(speaker_latent)
text_kv = get_kv_cache_text(text_input_ids)

t = random_sample_logit_normal_timestep_0_1(rng)

noise = torch.randn_like(latent)
x_noised = latent * (1 - t) + noise * t

v_pred = model(x_noised, t, text_mask, speaker_mask, text_kv, speaker_kv)

loss = ((v_pred - (noise - latent)) ** 2).mean()
```

These are both for the non-blockwise setup. For the blockwise-setup, it's probably best to rewrite the PyTorch model to process the entire latent and full set of block_indices (self-attention now is only within blocks, and latent cross-attention mask is computed based on block_ind_q > block_ind_latent; this approach is described briefly in https://jordandarefsky.com/blog/2025/echo under "(Dynamic-)Block-wise diffusion").

Here is the block_indices sampling code in JAX:

```
num_boundaries_uniform = jax.random.uniform(rng_num_boundaries, (x_noised.shape[0], ))
num_boundaries = 2 * (num_boundaries_uniform > 0.5).astype(jnp.int32) + 1 * (num_boundaries_uniform < 0.15).astype(jnp.int32)

MAX_NUM_BOUNDARIES = 2
segment_boundaries = jax.random.randint(rng_block, (x_noised.shape[0], MAX_NUM_BOUNDARIES), 0, x_noised.shape[1]) // (model_config.encoder_patch_size or 1) * (model_config.encoder_patch_size or 1)

block_indices = jnp.zeros((x_noised.shape[0], x_noised.shape[1]), dtype=jnp.int32)
for j in range(MAX_NUM_BOUNDARIES):
    block_indices += (jnp.arange(x_noised.shape[1])[None] >= segment_boundaries[:, j][:, None]) * (j < num_boundaries[..., None])
```

Sorry that I am not able to release something more turnkey at the moment.
